WIP - initial commit to the csv output

hufngvuowng opened the pull request.

Commit:
initial commit to the csv output

sbillinge commented on regolith/builders/recentcollabsbuilder.py:
```diff
@@ -104,23 +109,31 @@ def latex(self):
-            )
+                emp.sort(key=ene_date_key, reverse=True)
+                people_df = pd.DataFrame(ppl_names)
+                out_folder = "_build/recent-collabs/"
```
we should use regolith runcontrol to handle where outputs are placed.  I think it is rc.builddir

sbillinge commented on regolith/builders/recentcollabsbuilder.py:
```diff
@@ -104,23 +109,31 @@ def latex(self):
+                print(people_df.shape)
+                concat_df = pd.concat([head_part, people_df, tail_part], ignore_index=True)
+                concat_df.to_excel(out_folder + 'co.xslx')
+                break
```
what is this ``break`` doing?  Can we just delete it?

hufngvuowng replied:
Yeah I removed it. The `break` was added so that if it finds one instance of `p["_id"] == "sbillinge"` it'll stop. Earlier the code produced many dataframes and we thought this was due to the `for p in self.gtx["people"]:` loop. But I checked again and it produced only 1 dataframe even w/o the `break` so I guess it was some error in the code level earlier.

sbillinge commented on regolith/builders/recentcollabsbuilder.py:
```diff
@@ -104,23 +109,31 @@ def latex(self):
+                break
+
+
+
```
remove extra line

sbillinge commented on regolith/builders/recentcollabsbuilder.py:
```diff
@@ -104,23 +109,31 @@ def latex(self):
+                people_df.to_csv(out_folder + out_file, index=False)
+
+                excel_df = pd.read_excel('../templates/coa_template.xlsx')
+                head_part = excel_df.iloc[:49, :].copy()
```
what is the 49 doing here?  Do we want sthg like this hard coded in?

hufngvuowng replied:
49 is the row index in the excel file where the table of collaborators begins. Probably not something we want to hard code in, I'll try to figure that out.

sbillinge replied:
Thanks @tienhungvuong.  The point is that it shouldn't be hard-coded.  Create a module level variable called ``COAUTHOR_OFFSET``, for example and right at the top of the file, below the imports, assign the value of 49 to it, then get rid of the 49 down here.

sbillinge replied:
maybe ``COAUTHOR_TABLE_OFFSET`` is better actually

sbillinge reviewed (commented):
Good start, thanks guys!  I think the pandas approach will work well but better to use the regolith to handle the build location (see comment).

sbillinge commented:
in general it is usually bad design if you need a break, unless it is the
result of a conditional

On Thu, Feb 6, 2020 at 11:20 AM Hung Vuong <notifications@github.com> wrote:

> *@tienhungvuong* commented on this pull request.
> ------------------------------
>
> In regolith/builders/recentcollabsbuilder.py:
>
> > -            )
> +                emp.sort(key=ene_date_key, reverse=True)
> +                people_df = pd.DataFrame(ppl_names)
> +                out_folder = "_build/recent-collabs/"
> +                out_file = 'recent_collaborators.csv'
> +                people_df.to_csv(out_folder + out_file, index=False)
> +
> +                excel_df = pd.read_excel('../templates/coa_template.xlsx')
> +                head_part = excel_df.iloc[:49, :].copy()
> +                tail_part = excel_df.iloc[49:, :].copy()
> +                print(head_part.shape)
> +                print(tail_part.shape)
> +                print(people_df.shape)
> +                concat_df = pd.concat([head_part, people_df, tail_part], ignore_index=True)
> +                concat_df.to_excel(out_folder + 'co.xslx')
> +                break
>
> Yeah I removed it. The break was added so that if it finds one instance
> of p["_id"] == "sbillinge" it'll stop. Earlier the code produced many
> dataframes and we thought this was due to the for p in self.gtx["people"]:
> loop. But I checked again and it produced only 1 dataframe even w/o the
> break so I guess it was some error in the code level earlier.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/sbillinge/regolith/pull/14?email_source=notifications&email_token=ABAOWUIJWIF7UE7FCTV6FZTRBQ2GHA5CNFSM4KQUAI6KYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCURIRKY#discussion_r375936562>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABAOWUOQOM5PYB22UENHGQ3RBQ2GHANCNFSM4KQUAI6A>
> .
>


-- 
Professor Simon Billinge
Columbia University

Commit:
- Finished writing codes producing the excel file with the coauthors list appended.
- Next: use rc.builddir to make directory

hufngvuowng commented:
The excel file that the code produced lost all the formatting (bold text, hyperlinks, cell colors) in the template, so I don't know if it's a good idea to do it this way compared to manually pasting entries in the template. I can also try to see how to keep the formatting with python. Excel file attached below.
[coa_table.xlsx](https://github.com/sbillinge/regolith/files/4167029/coa_table.xlsx)

sbillinge commented:
The next thing to try would be to use jinja2.  Use the reimbursement
builder in regolith as the template.

On Thu, Feb 6, 2020 at 1:12 PM Hung Vuong <notifications@github.com> wrote:

> The excel file that the code produced lost all the formatting (bold text,
> hyperlinks, cell colors) in the template, so I don't know if it's a good
> idea to do it this way compared to manually pasting entries in the
> template. I can also try to see how to keep the formatting with python.
> Excel file attached below.
> coa_table.xlsx
> <https://github.com/sbillinge/regolith/files/4167029/coa_table.xlsx>
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/sbillinge/regolith/pull/14?email_source=notifications&email_token=ABAOWUM36BHWNWY6OEAA4YTRBRHIDA5CNFSM4KQUAI6KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOELAHELQ#issuecomment-583037486>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABAOWULE4X2TCXLABCYJD2LRBRHIDANCNFSM4KQUAI6A>
> .
>


-- 
Professor Simon Billinge
Columbia University

st3107 commented:
Here, I joint the discussion.

st3107 commented:
Actually, the reimbursement builder uses the "openpyxl" instead of "jinja2"

st3107 commented:
We are going to use the openpyxl. It will perserve the format. However, it seems that it is quite slow.

sbillinge commented:
🙂yes, super slow, but we will only build this occasionally and just for
one person usually so I think it is ok.

On Thu, Feb 6, 2020 at 2:55 PM Songsheng Tao <notifications@github.com>
wrote:

> We are going to use the openpyxl. It will perserve the format. However, it
> seems that it is quite slow.
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/sbillinge/regolith/pull/14?email_source=notifications&email_token=ABAOWUKDIPJLPMZBPNMTFF3RBRTLNA5CNFSM4KQUAI6KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOELAR4WY#issuecomment-583081563>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABAOWUKGRM6KOGEGIWOSPWLRBRTLNANCNFSM4KQUAI6A>
> .
>
-- 
Professor Simon Billinge
Columbia University

Commit:
- recentcollaboratorsbuilder.py (1) functionalized: turned simon's code into
a separate function that returns a list of people and institutions. turned
the csv/excel making codes into a separate function.
- coa_template.xlsx: modified the template: unmerged the cells.

sbillinge commented on regolith/builders/recentcollabsbuilder.py:
```diff
@@ -47,11 +54,26 @@ def construct_global_ctx(self):
 
-    def latex(self):
+    def get_ppl_inst_info(self, id, months):
+        '''
```
please use the numpy standard for docstrings

sbillinge commented on regolith/builders/recentcollabsbuilder.py:
```diff
@@ -47,11 +54,26 @@ def construct_global_ctx(self):
+        the person with the given id within the given number months from today
+        args:
+            - id: str, name of person of interest
+            - months: int, number of months
```
number of months what? since the last election? till my birthday?

sbillinge commented on regolith/builders/recentcollabsbuilder.py:
```diff
@@ -47,11 +54,26 @@ def construct_global_ctx(self):
+            - id: str, name of person of interest
+            - months: int, number of months
+        output: ppl_names, list of (c/a, people, institution, dept, last active)
+            - c/a: categories, either collaborator ('C') or co-author ('A'). here data is pulled
```
please work to make the explanations a bit clearer. Thx

sbillinge commented on regolith/builders/recentcollabsbuilder.py:
```diff
@@ -112,26 +134,49 @@ def latex(self):
+                cell.border = border
+        wb.save(''.join([out_folder, 'coa_table.xlsx']))
 
+    def latex(self):
```
What is this function for?  It seems to have things hard-coded. Is it used for testing?  If so, please remove as testing should be handled by the testing framework.

st3107 replied:
The 'latex' is the function that will be called to make the files. We will reformat the design like this:

'latex' do the taks by two steps: first, query the data from the database (which is a list of name and institution); second, use the data to generate csv and xlsx. Each step is done by a function (a 'get_data' function and a 'use_data' function).

sbillinge replied:
OK, I like that approach.  So we just need to remove the hard-coded sbillinge and 48.

The default behavior would normally be to loop over all people in people.yml, but override this behavior with the people list if one is specified at the command line.

for the 48, I would create a constant global variable at the top of the module with a helpful name, like ``LINES_TO_SKIP_TO_COLLABORATORS_TABLE_IN_EXCEL_FILE = 48`` and then put the constant name here.

sbillinge reviewed (commented):
Thanks The PR seems to contain a bunch of .ideas files.  Please clean these out and make sure that you have a global .gitignore so they are never accidentally committed again.  Maybe also add this to the .gitignore file for the project too to be on the safe side?

use ``git rm`` to remove them so they are removed from both file-system and to stop them being tracked by git (it is already too late that they are in the git commit history)

Please **never** use ``git add --all``  Be conscious of what you are adding to a commit and why.

This PR needs a regolith news and tests

Commit:
modified .gitignore: added .ideas

Commit:
removed .idea

Commit:
added .idea to the .gitignore file

Commit:
- reformatted the docstrings
- added more information on what each function does and returns

sbillinge commented on regolith/builders/recentcollabsbuilder.py:
```diff
@@ -55,21 +55,33 @@ def construct_global_ctx(self):
+        Parameters
+        ----------
+        id : str
+            id of the person of interest (i.e. "sbillinge")
```
numpy style would put a "the" at the beginning....so ``the id of the person..."

Also maybe clearer than "person of interest"  would be "person for whom the collaborator list is being built"

sbillinge commented on regolith/builders/recentcollabsbuilder.py:
```diff
@@ -55,21 +55,33 @@ def construct_global_ctx(self):
+        id : str
+            id of the person of interest (i.e. "sbillinge")
+        months : int
+            number of months from today
```
please fix the "the" issue everywhere.

For this I would maybe say "the number of months to go back in time when searching for jointly authored papers"

sbillinge commented on regolith/builders/recentcollabsbuilder.py:
```diff
@@ -55,21 +55,33 @@ def construct_global_ctx(self):
+                name
+            - institution : str
+                institution
+            - dept : str
```
please delete and remove from the signature if it is not being used yet.

sbillinge commented on regolith/builders/recentcollabsbuilder.py:
```diff
@@ -55,21 +55,33 @@ def construct_global_ctx(self):
+            - dept : str
+                additional info such as email/department to distinguish people with the same name
+                leave blank for now
+            - last active : str
```
please delete if it is not being used.

sbillinge commented on regolith/builders/recentcollabsbuilder.py:
```diff
@@ -139,15 +151,18 @@ def get_ppl_inst_info(self, id, months):
 
-    def make_csv_and_excel(self, id, months):
-        '''
+    def make_csv_and_excel(self, ppl_names):
```
Perhaps ``ppl`` is better than ``ppl_names`` since it is a list of tuples that contain more than the names

sbillinge commented on regolith/builders/recentcollabsbuilder.py:
```diff
@@ -139,15 +151,18 @@ def get_ppl_inst_info(self, id, months):
-        ppl_names = self.get_ppl_inst_info(id, months)
+        Parameters
+        ----------
+        ppl_names : list
```
put ``list of tuples`` here and remove from below.

sbillinge commented on regolith/builders/recentcollabsbuilder.py:
```diff
@@ -139,15 +151,18 @@ def get_ppl_inst_info(self, id, months):
+        Parameters
+        ----------
+        ppl_names : list
+            list of tuples. each tuple is a row to be added to the csv and the excel files
```
This should say what the thing is rather than its type, which is handled above, so ``"the collaborator information for the rows in the csv and excel files"``

sbillinge commented on regolith/builders/recentcollabsbuilder.py:
```diff
@@ -17,8 +20,16 @@
 from regolith.builders.basebuilder import LatexBuilderBase, latex_safe
 
-LATEX_OPTS = ["-halt-on-error", "-file-line-error"]
+# specifying border and font style
```
do we need this?  Doesn't jinja2 just inherit this from the template?

hufngvuowng replied:
actually 48 is the number of months from today to search from (searching for all collaborators who worked with "sbillinge" within 48 months from today). There's a global variable for the lines to skip to the collaborators table in the excel file, `COAUTHOR_TABLE_OFFSET`, declared at the beginning of the code.

sbillinge commented on regolith/builders/recentcollabsbuilder.py:
```diff
@@ -2,6 +2,9 @@
 import datetime as dt
 from copy import copy
+from typing import List, Any, Tuple
+import pandas as pd
```
can we remove pandas?  I am not convinced it is adding sufficient value here to (a) change the requirements of regolith and (b) introduce a new datastructure (panda's dataframe).  If you can convince me otherwise we can maybe justify it, but right now I don't see it.

st3107 replied:
@sbillinge There are three options to parse and save the data in .csv file: (1) panadas; (2) built-in csv module; (3) jinja2.

(1) Pro: just one line of code (pd.DataFrame.to_csv), quick; Con: need install pandas module; may be overkill

Code:

pd.DataFrame([...]).to_csv('...')

(2) Pro: it is a build in model in python, quick; Con: four lines of code

Code:

with open('xxx.csv', "r") as f:
         w = csv.writer(f)
         w.write([...])
         for row in rows:
               w.write(row)

(3) Pro: do not need to worry about the build of directroy; Con: need templates; need to reformat the data into a list of string with "," delimeters.

Code:

rows = [','.join(tup) for tup in tups]
self.render('xx.csv', 'xx.csv', rows)

Template:

{% for row in rows %}
{{row}}

Pandas may be not necessary here. However, I am going to introduce the pandas in the requirements in regolith for my "beamplanbuilder" module bceaseuse it really needs that to generate the table in the latex form. I think it is ok to use pandas here to make the code simple.

st3107 replied:
I forgot to mention that the Jinja2 is not capable to do this task.

For example, if there is "," in a human name, jinja2 cannot identify that it is a data instead of a delimeter and it will break the .csv format.

So our options are (1) pandas Or (2) csv

sbillinge reviewed (commented):
Thanks guys.  please see my comments

sbillinge commented:
doesn't this actually break the csv in any case?  For example,
Last, First is an ok entry for excel but not for csv, right?

On Sat, Feb 8, 2020 at 3:59 PM Songsheng Tao <notifications@github.com>
wrote:

> *@st3107* commented on this pull request.
> ------------------------------
>
> In regolith/builders/recentcollabsbuilder.py:
>
> > @@ -2,6 +2,9 @@
>  import os
>  import datetime as dt
>  from copy import copy
> +from typing import List, Any, Tuple
> +import pandas as pd
>
> I forgot to mention that the Jinja2 is not capable to do this task.
>
> For example, if there is "," in a human name, jinja2 cannot identify that
> it is a data instead of a delimeter and it will break the .csv format.
>
> So our options are (1) pandas Or (2) csv
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/sbillinge/regolith/pull/14?email_source=notifications&email_token=ABAOWUK77AQBYXU35G4LRDDRB4MKHA5CNFSM4KQUAI6KYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCUYYYUQ#discussion_r376734025>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABAOWUNTQDUJOBFMGBGN5VLRB4MKHANCNFSM4KQUAI6A>
> .
>


-- 
Professor Simon Billinge
Columbia University

st3107 commented:
The excel is buily by openpyxl not the jinja2. And the openpyxl set the value of cell one by one so it doesn't care about if there is a comma in row or not since it doesn't operate on row.

sbillinge commented:
my point is not what is building things.  It is that the "lastname,
firstname" format that NSF wants will work for excel but not for csv in any
case.  So we probably shouldn't build csv at all maybe?

S

On Sat, Feb 8, 2020 at 4:29 PM Songsheng Tao <notifications@github.com>
wrote:

> The excel is buily by openpyxl not the jinja2. And the openpyxl set the
> value of cell one by one so it doesn't care about if there is a comma in
> row or not since it doesn't operate on row.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/sbillinge/regolith/pull/14?email_source=notifications&email_token=ABAOWUMXYWTLDP4VAD4O6CLRB4PZPA5CNFSM4KQUAI6KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOELF4CVI#issuecomment-583778645>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABAOWUOIUUJE76I7MAUW473RB4PZPANCNFSM4KQUAI6A>
> .
>


-- 
Professor Simon Billinge
Columbia University

hufngvuowng commented:
> my point is not what is building things. It is that the "lastname, firstname" format that NSF wants will work for excel but not for csv in any case. So we probably shouldn't build csv at all maybe? S
> […](#)
> On Sat, Feb 8, 2020 at 4:29 PM Songsheng Tao ***@***.***> wrote: The excel is buily by openpyxl not the jinja2. And the openpyxl set the value of cell one by one so it doesn't care about if there is a comma in row or not since it doesn't operate on row. — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub <#14?email_source=notifications&email_token=ABAOWUMXYWTLDP4VAD4O6CLRB4PZPA5CNFSM4KQUAI6KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOELF4CVI#issuecomment-583778645>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ABAOWUOIUUJE76I7MAUW473RB4PZPANCNFSM4KQUAI6A> .
> -- Professor Simon Billinge Columbia University

we just used pandas because it produces the csv file in 1 line of code. if there's no use for the csv file then i think pandas is no longer necessary. i'll remove pandas and fix the stuff according to your other comments.

sbillinge commented:
....so, how do you suggest moving forward?

On Mon, Feb 10, 2020 at 11:36 AM Hung Vuong <notifications@github.com>
wrote:

> *@tienhungvuong* commented on this pull request.
> ------------------------------
>
> In regolith/builders/recentcollabsbuilder.py:
>
> > @@ -17,8 +20,16 @@
>  from regolith.sorters import doc_date_key, ene_date_key, position_key
>  from regolith.builders.basebuilder import LatexBuilderBase, latex_safe
>
> -LATEX_OPTS = ["-halt-on-error", "-file-line-error"]
> +# specifying border and font style
>
> actually 48 is the number of months from today to search from (searching
> for all collaborators who worked with "sbillinge" within 48 months from
> today). There's a global variable for the lines to skip to the
> collaborators table in the excel file, COAUTHOR_TABLE_OFFSET, declared at
> the beginning of the code.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/sbillinge/regolith/pull/14?email_source=notifications&email_token=ABAOWUOJ2S76TQXPTRYW4TDRCF7AHA5CNFSM4KQUAI6KYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCU4W5AY#discussion_r377176791>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABAOWULNWSZ52TUFAPDKC2TRCF7AHANCNFSM4KQUAI6A>
> .
>


-- 
Professor Simon Billinge
Columbia University

hufngvuowng commented:
I think we can specify global variables at the top of the code for the ID and number of months like `id = "sbillinge"` and  `num_months = 48`, and pass those into the `latex` function. Like if anyone wants to use different id and number of months they can modify those variables at the top without having to go into the `latex` function to do so.

sbillinge commented:
Nailed it, thanks!  Actually, this is a good lesson why declaring
descriptive variables is always better than hard coding numbers....I mixed
up the 48 and the 49...

Global constants should be capitalized, so it would NUM_MONTHS.  I think we
can take id from the CLI, i.e., read it from self.rc.client.people, or
whatever it is called.

S

On Mon, Feb 10, 2020 at 11:58 AM Hung Vuong <notifications@github.com>
wrote:

> I think we can specify global variables at the top of the code for the ID
> and number of months like id = "sbillinge" and num_months = 48, and pass
> those into the latex function. Like if anyone wants to use different id
> and number of months they can modify those variables at the top without
> having to go into the latex function to do so.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/sbillinge/regolith/pull/14?email_source=notifications&email_token=ABAOWUMWGIPI6SITOHPSKOTRCGBTRA5CNFSM4KQUAI6KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOELJIUUA#issuecomment-584223312>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABAOWUMJD3YIPKNGEUML7GLRCGBTRANCNFSM4KQUAI6A>
> .
>


-- 
Professor Simon Billinge
Columbia University

Commit:
- added global variables at the top of the code
- modified docstrings of functions
- used rc.builddir to make directory

sbillinge commented:
@tienhungvuong I am ready to merge this but it has conflicts.  Please could you resolve them with a rebase?

Thanks,

S

hufngvuowng commented:
Closing this since I already redid the COA template table.
